README.rst
cloudns_api
###########

A Python 3 interface to the ClouDNS.net API
===========================================

This project is an independent project not developed by
`CloudNS.net <https://cloudns.net>`__.

The interface is designed to be simple and intuitive. Entities are represented
as modules. Each has a list, create, update, and delete function (when
appropriate, ie, you cannot delete an SOA record). Some entities have functions
specific to them, but whenever possible, the arguments are consistent. One of
the major goals of this project is code readability.

Another major goal is for the code to have 90%+ test coverage.

For documentation on the `CloudNS.net <https://cloudns.net>`__ API see `here
<https://www.cloudns.net/wiki/article/41/>`__.

Please submit any bug reports and bug fixes on github `here
<https://github.com/prestix-studio/>`__.



Installing and including in projects
====================================


Installing cloudns_api
----------------------

.. code:: bash

    $ pip install cloudns_api


In order to authenticate, you must first generate a CloudNS auth user or sub
user and password combination. (See `here
<https://www.cloudns.net/wiki/article/42/>`__ for instructions.) Then you must
set the appropriate username and password values as environment variables as
follows:

.. code:: bash

    # USERNAME:
    export CLOUDNS_API_AUTH_ID=my_user
    #      - or -
    export CLOUDNS_API_SUB_AUTH_ID=my_user
    #      - or -
    export CLOUDNS_API_SUB_AUTH_USER=my_user

    # PASSWORD:
    # This should be the password that corresponds to the above user or sub
    # user.
    export CLOUDNS_API_AUTH_PASSWORD=my_password


When you are debugging, you can set the environment variable
`CLOUDNS_API_DEBUG` to True:

.. code:: bash

    export CLOUDNS_API_DEBUG=True


To make things easier, you could put these in your python virtual environment
or use a package like
`python-dotenv <https://github.com/theskumar/python-dotenv>`__ to automatically
load your environment variables. Be sure to *not* include your private username
and password in your public repositories.


Running Tests
-------------

.. code:: bash

    $ cd <cloudns_api directory>
    $ py.test

    $ py.test -x                    # Stop on failures
    $ py.test tests/test_record.py  # Specific test
    $ py.test -k validate           # Run tests with 'validate' in the name


Importing and Basic Usage
-------------------------

.. code:: python

    >>> import cloudns_api

    >>> response = cloudns_api.zone.list(search='example')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      [
                {
                    "name":     "example.com",
                    "type":     "master",
                    "zone":     "domain",
                    "status":   "1"
                },
                {
                    "name":     "example.net",
                    "type":     "master",
                    "zone":     "domain",
                    "status":   "1"
                },
                {
                    "name":     "example.org",
                    "type":     "master",
                    "zone":     "domain",
                    "status":   "1"
                }
            ]
        }



API Reference
=============

Introduction
------------

We have created the API to be consistent and predictable. API calls usually
include `list`, `create`, `get`, and `update` functions that do exactly what
they say. Arguments are passed in a consistent manner across all functions.
When an argument accepts an integer, it can be passed as an integer or a string
of that integer.

All arguments are passed to the API functions as keyword arguments. Each API
function turns these arguments into a `Parameters` instance. The construction
of the `Parameters` instance can also include information for validating the
arguments. This validation happens by default when the object is instantiated.
If a validation error occurs, the exception is thrown and handled in the api
decorator. A parameter is required unless the optional flag is set to `True`.
The name of the parameter is matched to a validation function unless the
`validate_as` option is set on that particular parameter. The `Parameters`
object has a to_dict() method that returns the parameters as a key-value dict
to be passed on to the CloudNS API using requests.

The CloudNS API sometimes uses camel case and sometimes uses dashes in its
parameters. In our API, we convert both of these to snake case for consistency
and in order to be "pythonic".

API `update` functions require all required parameters to be passed. This can
be inconvenient at times, so cloudns_api includes an argument `patch` that when
set to `True` allows you to pass only arguments you wish to change. Behind the
scenes, the API will get the existing data and merge it with the new data for
the update call. We've also included the convenient `patch` function as a
wrapper around `update` with the `patch` argument set to `True`.

The cloudns_api includes these two helpful functions for checking your login
credentials and retrieving your CloudNS nameservers:

.. code:: python

    >>> print(cloudns_api.api.get_login())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "status":             "Success",
                "status_description": "Success login."
            }
        }

    >>> print(cloudns_api.api.get_nameservers())


ApiResponse
^^^^^^^^^^^

All API calls return an `ApiResponse` instance. The `ApiResponse` object is a
wrapper around the basic response object from the `requests
<https://github.com/kennethreitz/requests>`__ library. We've added a few
properties and functions specific to our needs here.

.. code:: python

    >>> print(response.success)      # See if a response succeeded

    >>> print(response.status_code)  # Get the status of a response

    >>> print(response.payload)      # The payload of the response
                                     # Note that the parameters are converted
                                     # to camel case here.

    >>> print(response.json())       # Get the response as json object

    >>> print(response.string())     # Get the response as json string


ApiParameter
^^^^^^^^^^^^

The `ApiParameter` object is responsible for describing the kinds of parameters
to pass to the api function and how these parameters should be validated.
Understanding the ApiParameter object is not necessary for using the API, but
can be helpful to see what is going on under the hood.

By default, an ApiParameter validates its parameters upon initialization. But
if the `validate` parameter is set to false, this can be deferred until later.
You can then call the `validate()` method to manually validate the parameters.

.. code:: python

    >>> print(parameters.validate())  # Validates the parameters according to
                                      # their definitions

    >>> print(response.to_dict())  # Returns the parameters as a dict. Used
                                   # when passing the parameters to requests.

A full discription of how an ApiParameter object and its parameter definition
works can be found in the `cloudns_api/parameters.py` file. You may also need
to reference the `cloudns_api/validation.py` module to see how validation
works.


DNS ZONE
--------

Parameters:

+ page - int/string (optional) Page number to show.
+ rows_per_page - int/string (optional) Number of rows per page to show.
+ search - string (optional) Optional string to filter results by.
+ group_id - int/string (optional) Optional group id to filter results by.

Response Parameters:

+ name - Domain name.
+ type - Zone type (Master, Slave, Parked, GeoDNS)
+ zone -
+ status - active (1) or inactive(0)


Listing DNS Zones
^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.list(search='example')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      [
                {
                    "name":     "example.com",
                    "type":     "master",
                    "zone":     "domain",
                    "status":   "1"
                },
                {
                    "name":     "example.net",
                    "type":     "master",
                    "zone":     "domain",
                    "status":   "1"
                },
                {
                    "name":     "example.org",
                    "type":     "master",
                    "zone":     "domain",
                    "status":   "1"
                }
            ]
        }

    >>> print(cloudns_api.zone.get_page_count(rows_per_page=10))  # Get page count


Creating DNS Zones
^^^^^^^^^^^^^^^^^^

NOTE: The nameserver argument doesn't seem to currently work on ClouDNS's
servers.

.. code:: python

    >>> response = cloudns_api.zone.create(domain_name='example.com',
                                           zone_type='master')
    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "status": "Success",
                "status_description":
                    "Domain zone example.com was created successfully."
            }
        }


Getting a DNS Zone
^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.get(domain_name='example.com')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "name":     "example.com",
                "type":     "master",
                "zone":     "domain",
                "status":   "1"
            }
        }


Updating a DNS Zone Serial Number
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.update(domain_name='example.com')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "status": "Success",
                "status_description":
                    "Domain zone example.com was updated successfully."
            }
        }


Activating/Deactivating a DNS Zone
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.activate(domain_name='example.com')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "status": "Success",
                "status_description":
                    "The zone was activated!"
            }
        }

    >>> cloudns_api.zone.deactivate(domain_name='example.com')

    >>> cloudns_api.zone.toggle_activation(domain_name='example.com')


Deleting a DNS Zone
^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.delete(domain_name='example.com')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "status": "Success",
                "status_description":
                    "Domain zone example.com was deleted successfully."
            }
        }


Getting ClouDNS Zone Stats
^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.get_stats()

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "count": "25",  # Number of zones used
                "limit": "40"   # Number of zones allowed by your plan
            }
        }


Check if DNSSEC is available for a zone
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.dnssec_available(domain_name='example.com')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      1,
        }


Activating/Deactivating a DNSSEC
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.dnssec_activate(domain_name='example.com')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "status": "Success",
                "status_description":
                    "The DNSSEC is activated for your zone. The keys will be generated soon."
            }
        }

    >>> cloudns_api.zone.deactivate(domain_name='example.com')


Getting DNSSEC DS Records
^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.dnssec_ds_records(domain_name='example.com')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
             "payload": {
                "status": "1",
                "ds": [
                    "example.com. 3600 IN DS 9813 13 2 613FDE9D90DB360EE4DDC1E18170D3306147A95E4F77177017C83E31057B9141"
                ],
                "ds_records": [
                    {
                        "digest": "613FDE9D90DB360EE4DDC1E18170D3306147A95E4F77177017C83E31057B9141",
                        "key_tag": "9813",
                        "algorithm": "13",
                        "algorithm_name": "ECDSA SHA-256",
                        "digest_type": "2",
                        "digest_type_name": "SHA-256"
                    }
                ],
                "dnskey": [
                    "example.com. 3600 IN DNSKEY 257 3 13 tDYgHxnS3cbLb9B2B2l+SsawWiG4jOzoFmnjy7PVL0NK5qiil/254sZLxEhXs0LNiL6YxcRVzYdHLkWi074SuQ==",
                    "example.com. 3600 IN DNSKEY 256 3 13 Nr9P1PdBNRCI7mpF7Nrx72rNZ7EQcHlVggUBJR0E9l+W0j37WlpluKM4qv/WVn/QsZxQOU1eSMPPyIXlT3sCvw=="
                ]
            }
        }


Checking if a DNS Zone Has Been Updated on all Servers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.zone.is_updated(domain_name='example.com')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      True
        }


SOA Record
----------

Every domain zone contains one SOA record that contains the current version of
the data in the zone, the administrator of the zone record, and TTL information
for the zone.

These functions only work for master zones.

SOA Parameters:

+ domain_name - string (required) Domain name or reverse zone name whose SOA
  details you want to modify.
+ primary_ns - string (required) Hostname of primary nameserver.
+ admin_mail - string (required) DNS zone administrator's e-mail.
+ refresh - integer (required) The time in seconds that a secondary DNS server
  waits before querying the primary DNS server's SOA record to check for
  changes. Rate can be any integer from 1200 to 43200 seconds.
+ retry - integer (required) The time in seconds that a secondary server waits
  before retrying a failed zone transfer. Usually, the retry rate is less than
  the refresh rate. Rate can be any integer from 180 to 2419200 seconds.
+ expire - integer (required) The time in seconds that a secondary server will
  keep trying to complete a zone transfer. If this time expires before a
  successful zone transfer, the secondary server will expire its zone file. The
  secondary will stop answering queries, as it considers its data too old to be
  reliable. Time can be any integer from 1209600 to 2419200 seconds.
+ default_ttl - integer (required) The minimum time-to-live value applies to
  all resource records in the zone file. TTL can be any integer from 60 to
  2419200 seconds.

Note that ClouDNS automatically increments the serial number when the zone is
updated or changed.


Getting the SOA for a domain
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.soa.get('example.com')

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "admin_mail":     "admin@example.com",
                "default_ttl":    "3600",
                "expire":         "1209600",
                "primary_ns":     "ns1.example.com",
                "refresh":        "7200",
                "retry":          "1800",
                "serial_number":  "2019060601"
            }
        }


Updating the SOA for a domain
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: python

    >>> response = cloudns_api.soa.update(
            'example.com',  # The domain to patch
            admin_mail='admin@example.com',
            default_ttl=3600,
            expire=1209600,
            primary_ns='ns1.example.com',
            refresh=7200,
            retry=1800,
        )

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "status": "Success",
                "status_description":
                    "The SOA record was modified successfully."
            }
        }


Patch Updating the SOA for a domain
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A patch update allows you to specify only the parameters you wish to change.

.. code:: python

    >>> response = cloudns_api.soa.patch(
            'example.com',  # The domain to patch
            admin_mail='admin@example.com',
            primary_ns='ns1.example.com',
        )

    >>> print(response.json())

        {
            "success":      True,
            "status_code":  200,
            "payload":      {
                "status": "Success",
                "status_description":
                    "The SOA record was modified successfully."
            }
        }


DNS Records
-----------

Besides the SOA record, a domain can have a number of other records.

+ A record - points a hostname to an IPv4 address.

  * host - subdomain to optionally add to main domain

  * record - an IPv4

  * ttl - time to keep record in cache


+ AAAA record - points a hostname to an IPv6 address.

  * host - subdomain to optionally add to main domain

  * record - an IPv6

  * ttl - time to keep record in cache



CHANGELOG.rst
.. :changelog:

Release History
---------------

0.9.6 (Jul 14, 2021)
++++++++++++++++++++

* Fixes: Fixes issue where python-dotenv was ignoring `.env` config.


0.9.5 (May 17, 2021)
++++++++++++++++++++

* Fixes: Adds LICENSE to binary and source distributions.


0.9.4 (May 17, 2021)
++++++++++++++++++++

* Updates copyright date.
* Adds LICENSE to binary and source distributions.


0.9.3 (Jun 6, 2020)
+++++++++++++++++++

* Increments version after the previous version to revert back to preventing
  installation on Python 2.


0.9.2 (Jun 6, 2020)
+++++++++++++++++++

* Released with a better way to warn Python 2 users that this library only
  works in Python 3. The previous version failed with a confusing error
  message.


0.9.1 (Jun 6, 2020)
+++++++++++++++++++

* Adds notice that this library must be run with Python 3.


0.9 (May 5, 2020)
+++++++++++++++++

* Version 1 Release candidate.
* Adds support for checking if a DNS zone is updated (`is_updated`).
* Updates error handling in `publish` script.


0.8 (April 2, 2020)
+++++++++++++++++++

* Fourth Beta Release.
* Adds support for TLSA records.
* Minor fixes and tweaks to documentation for clarification.
* Fixes a bug where tests required auth environment variables to be set in
  order to run successfully.


0.7 (November 16, 2019)
+++++++++++++++++++++++

* Third Beta Release.
* Adds DNS Support for zones (Thanks to
  `kristinn <https://github.com/kristinn>`__).
* Allows authentication with sub-users through environment variables
  'CLOUDNS_API_SUB_AUTH_ID' or 'CLOUDNS_API_SUB_AUTH_USER'. (Thanks to
  `sadimusi <https://github.com/sadimusi>`__).


0.6 (September 14, 2019)
++++++++++++++++++++++++

* Second Beta Release.
* Corrects wrong bash variable 'CLOUDNS_API_AUTH_USER' to 'CLOUDNS_API_AUTH_ID'
  in README.rst
* Changes the way errors are thrown when auth environment variables are not
  set. This allows the library to be included without these variables without
  immediately throwing an error. Now the error is only thrown when they are
  needed to authenticate.


0.5 (July 13, 2019)
+++++++++++++++++++

* Beta Release.
* Updates github address to Prestix Studio, LLC github account.
* Corrects pypi reference.
* Adds publish script to push next version to pypi automatically.
* Minor code cleanup.


0.4 (July 07, 2019)
+++++++++++++++++++

* Alpha Release - all mvp features.
* Minor bugfixes from v0.3.


0.3 (June 05, 2019)
+++++++++++++++++++

* Planning/Testing Release 'b' - updates to core modules.


0.2 (September 19, 2018)
++++++++++++++++++++++++

* Planning/Testing Release 'a' - adds some real code for initial testing.


0.1 (September 16, 2018)
++++++++++++++++++++++++

* Initial Planning Release to secure 'cloudns-api' name.


cloudns_api/validation.py
# -*- coding: utf-8 -*-
#
# name:             validation.py
# author:           Harold Bradley III | Prestix Studio, LLC.
# email:            harold@prestix.studio
# created on:       09/16/2018
#

"""
cloudns_api.validation
~~~~~~~~~~~~~~~~~~~~~~

This module contains functions to validate fields passed into the API.
"""

import re


class ValidationError(Exception):
    """Exception thrown when a validation error has occured."""
    def __init__(self, fieldname, message, *args, **kwargs):
        """Initialize ValidationError with `fieldname` and `message` values."""
        self.details = {
            'fieldname': fieldname,
            'message':   message,
        }
        super(ValidationError, self).__init__(message, *args, **kwargs)

    def get_details(self):
        """Returns a list of error details."""
        return [self.details]


class ValidationErrorsBatch(ValidationError):
    """Exception thrown when multiple validation errors have occured."""
    def __init__(self, validation_errors, *args, **kwargs):
        """Initialize ValidationError with `validation_errors` list."""
        self.validation_errors = validation_errors
        super(ValidationError, self).__init__('Validation errors occured.',
                                              *args, **kwargs)

    def get_details(self):
        """Returns a list of error details."""
        return [error.details for error in self.validation_errors]


def validate(fieldname, value, optional=False, validate_as=None, *args,
             **kwargs):
    """Validates a value for a particular fieldtype.
    Returns True if it is valid; raises an error otherwise.

    :param fieldname: string, the fieldname to validate (determines the type of
        validation to use)
    :param value: mixed, the value to validate
    :param optional: bool, if True, accept None as a valid value
    :param validate_as: string, an optional argument to specifically select a
        validation function to run. If this is not passed, the fieldname will
        be used.
    """
    if optional and not value:
        return True
    elif value is None or value == '':
        raise ValidationError(fieldname, 'This field (' + fieldname +
                              ') is required.')

    if not validate_as:
        validate_as = fieldname

    if validate_as not in validation_functions:
        return True

    if validation_functions[validate_as](value, fieldname, *args, **kwargs):
        return True

    # If for some reason the validation fails, but no error was raised,
    # raise one here. This should hopefully never happen
    raise ValidationError('Unexpected validation error.')  # pragma: no cover


##
# Specific Validation Functions

def is_algorithm(value, fieldname='algorithm', **kwargs):
    """Validates and returns True if the value is a proper algorithm.
    Otherwise, raises a validation error."""
    try:
        if value.upper() not in ALGORITHMS:
            raise ValidationError(fieldname,
                                  'This field must be RSA, DSA, ECDSA, or ' +
                                  'Ed25519.')
    # If value isn't a string, upper() raises AttributeError
    except AttributeError:
        if value not in [1, 2, 3, 4]:
            raise ValidationError(fieldname,
                                  'This field must be RSA, DSA, ECDSA, or ' +
                                  'Ed25519.')
    return True


ALGORITHMS = ['RSA', 'DSA', 'ECDSA', 'ED25519']


def is_api_bool(value, fieldname='api-bool', **kwargs):
    """Validates and returns True if the value is a 0 or 1. Otherwise, raises a
    validation error."""
    if value != 0 and value != 1:
        raise ValidationError(fieldname,
                              'This field must be 0 or 1.')
    return True


def is_caa_flag(value, fieldname='caa-flag', **kwargs):
    """Validates and returns True if the value is 0 or 128. Otherwise, raises a
    validation error."""
    if value != 0 and value != 128:
        raise ValidationError(fieldname,
                              'This field must be 0 (non-critical) or 128 ' +
                              '(critical).')
    return True


def is_caa_type(value, fieldname='caa-type', **kwargs):
    """Validates and returns True if value is a caa type. Otherwise, raises a
    validation error."""
    try:
        if value.lower() not in CAA_TYPES:
            raise ValidationError(fieldname,
                                  'This field must be one of issue, ' +
                                  'issuewild, iodef.')
    # If value isn't a string, lower() raises AttributeError
    except AttributeError:
        raise ValidationError(fieldname,
                              'This field must be one of issue, ' +
                              'issuewild, iodef.')
    return True


CAA_TYPES = ['issue', 'issuewild', 'iodef']


def is_domain_name(value, fieldname='domain-name', **kwargs):
    """Validates and returns True if the value is a valid domain name.
    Otherwise, raises a validation error."""
    if not re.match(
       r'^((?=[a-z0-9-]{1,63}\.)(xn--)?[a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,63}$',
       value):
        raise ValidationError(fieldname,
                              'This field must be a valid domain name.')
    return True


def is_email(value, fieldname='email', **kwargs):
    """Validates and returns true if the value is a valid email. Otherwise,
    raises a validation error."""
    if not re.match(r'(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)',
                    value):
        raise ValidationError(fieldname,
                              'This field must be a valid email.')
    return True


def is_fptype(value, fieldname='fptype', **kwargs):
    """Validates and returns True if the value is a proper fingerprint type.
    Otherwise, raises a validation error."""
    try:
        if value.upper() not in FP_TYPES:
            raise ValidationError(fieldname,
                                  'This field must be one of SHA-1 or ' +
                                  'SHA-256.')
    # If value isn't a string, upper() raises AttributeError
    except AttributeError:
        if value not in [1, 2]:
            raise ValidationError(fieldname,
                                  'This field must be one of SHA-1 or ' +
                                  'SHA-256.')
    return True


FP_TYPES = ['SHA-1', 'SHA-256']


def is_int(value, fieldname='int', min_value=None, max_value=None, **kwargs):
    """Validates and returns True if the value is an integer (within
    min_value/max_value range). Otherwise, raises a validation error."""
    try:
        value += 0  # Try it to see if it is an integer
    except TypeError:
        if re.findall('[^0-9]', value):
            raise ValidationError(fieldname, 'This field must be an integer.')

    if min_value and int(value) < min_value:
        raise ValidationError(fieldname,
                              'This field must be greater than ' +
                              str(min_value) + '.')
    if max_value and int(value) > max_value:
        raise ValidationError(fieldname,
                              'This field must be less than ' +
                              str(max_value) + '.')
    return True


def is_ipv4(value, fieldname='ipv4', **kwargs):
    """Validates and returns True if the value is an IPv4 address. Otherwise,
    raises a validation error. Note that this is just a rudimentary check."""
    octets = []

    if hasattr(value, 'split'):
        octets = [o for o in value.split('.') if o.isdigit() and
                  0 <= int(o) and int(o) <= 255]

    if len(octets) != 4:
        raise ValidationError(fieldname,
                              'This field must be a valid IPv4 address.')

    return True


def is_ipv6(value, fieldname='ipv6', **kwargs):
    """Validates and returns True if the value is an IPv6 address. Otherwise,
    raises a validation error. Note that this is just a rudimentary check."""
    hextet = []

    if hasattr(value, 'split'):
        try:
            hextet = [h for h in value.split(':')
                      if 0 <= int(h, 16) and int(h, 16) <= 65535]
        except ValueError:
            hextet = []

    if len(hextet) != 8:
        raise ValidationError(fieldname,
                              'This field must be a valid IPv6 address.')

    return True


def is_record_type(value, fieldname='record-type', **kwargs):
    """Validates and returns True if the value is a valid domain record type.
    Otherwise, raises a validation error."""
    try:
        if value.upper() not in RECORD_TYPES:
            raise ValidationError(fieldname,
                                  'This field must be a valid domain record' +
                                  ' type.')
    # If value isn't a string, upper() raises AttributeError
    except AttributeError:
        raise ValidationError(fieldname,
                              'This field must be a valid domain record type.')
    return True


RECORD_TYPES = ['A', 'AAAA', 'MX', 'CNAME', 'TXT', 'SPF', 'NS', 'SRV', 'WR',
                'RP', 'SSHFP', 'ALIAS', 'CAA', 'NAPTR', 'PTR']


def is_redirect_type(value, fieldname='redirect-type', **kwargs):
    """Validates and returns True if the value is 301 or 302. Otherwise, raises
    a validation error."""
    if value != 301 and value != 302:
        raise ValidationError(fieldname,
                              'This field must be 301 (permanent) or 302 ' +
                              '(temporary).')
    return True


def is_required(value, fieldname='required', **kwargs):
    """Validates and returns True if any value is provided. Otherwise, raises a
    validation error."""
    if value is None or value == '':
        raise ValidationError(fieldname,
                              'This field is required.')
    return True


def is_rows_per_page(value, fieldname='rows-per-page', **kwargs):
    """Validates and returns True if a proper rows-per-page value is provided
    (10, 20, 30, 50, or 100). Otherwise, raises a valiadation error."""
    if value not in [10, 20, 30, 50, 100, '10', '20', '30', '50', '100']:
        raise ValidationError(fieldname,
                              'This field must be one of: 10, 20, 30, 50, or' +
                              ' 100.')
    return True


def is_tlsa_matching_type(value, fieldname='tlsa_matching_type', **kwargs):
    """Validates and returns True if a proper tlsa_matching_type value is
    provided (0, 1, 2). Otherwise, raises a valiadation error."""
    if value not in [0, 1, 2, '0', '1', '2']:
        raise ValidationError(fieldname,
                              'This field must be one of: 0, 1, or 2')
    return True


def is_tlsa_selector(value, fieldname='tlsa_selector', **kwargs):
    """Validates and returns True if a proper tlsa_selector value is provided
    (0, 1). Otherwise, raises a valiadation error."""
    if value not in [0, 1, '0', '1']:
        raise ValidationError(fieldname,
                              'This field must be one of: 0 or 1')
    return True


def is_tlsa_usage(value, fieldname='tlsa_usage', **kwargs):
    """Validates and returns True if a proper tlsa_usage value is provided (0,
    1, 2, 3). Otherwise, raises a valiadation error."""
    if value not in [0, 1, 2, 3, '0', '1', '2', '3']:
        raise ValidationError(fieldname,
                              'This field must be one of: 0, 1, 2, or 3')
    return True


def is_ttl(value, fieldname='ttl', **kwargs):
    """Validates and returns True if the value is a valid ClouDNS ttl.
    Otherwise, raises a validation error."""
    if hasattr(value, 'lower'):
        value = value.lower()

    if value not in TTLS and value not in TTL_STRINGS:
        raise ValidationError(fieldname, 'This field must be a valid ttl. ' +
                              '(1 minute, 5 minutes, 15 minutes, 30 minutes,' +
                              ' 1 hour, 6 hours, 12 hours, 1 day, 2 days, 3 ' +
                              'days, 1 week, 2 weeks, or 1 month) or (60, ' +
                              '300, 900, 1800, 3600, 21600, 43200, 86400, ' +
                              '172800, 259200, 604800, 1209600, or 2592000)')
    return True


TTL_STRINGS = ['1 minute', '5 minutes', '15 minutes', '30 minutes', '1 hour',
               '6 hours', '12 hours', '1 day', '2 days', '3 days', '1 week',
               '2 weeks', '1 month', '60', '300', '900', '1800', '3600',
               '21600', '43200', '86400', '172800', '259200', '604800',
               '1209600', '2592000']

TTLS = [60, 300, 900, 1800, 3600, 21600, 43200, 86400, 172800, 259200, 604800,
        1209600, 2592000]


def is_valid(value, fieldname='valid', **kwargs):
    """This is a stub. It always returns true."""
    return True


def is_zone_type(value, fieldname='zone-type', **kwargs):
    """Validates and returns True if the value is a valid zone type.
    Otherwise, raises a validation error."""
    if hasattr(value, 'lower'):
        value = value.lower()

    if value not in ZONE_TYPES:
        raise ValidationError(fieldname, 'This field must be a valid zone ' +
                              'type. (master, slave, parked, or geodns)')
    return True


ZONE_TYPES = ['master', 'slave', 'parked', 'geodns', 'domain', 'reverse']


# Set up validation functions dict
validation_functions = {
    'admin-mail':          is_email,
    'algorithm':           is_algorithm,
    'bool':                is_api_bool,
    'caa_flag':            is_caa_flag,
    'caa_type':            is_caa_type,
    'caa_value':           is_valid,
    'default-ttl':         is_ttl,
    'domain-name':         is_domain_name,
    'email':               is_email,
    'fptype':              is_fptype,
    'frame':               is_api_bool,
    'frame-title':         is_valid,
    'geodns-location':     is_int,
    'integer':             is_int,
    'ipv4':                is_ipv4,
    'ipv6':                is_ipv6,
    'mail':                is_email,
    'page':                is_int,
    'port':                is_int,
    'primary-ns':          is_domain_name,
    'priority':            is_int,
    'record':              is_required,
    'record-id':           is_int,
    'record-type':         is_record_type,
    'redirect-type':       is_redirect_type,
    'refresh':             is_int,
    'required':            is_required,
    'rows-per-page':       is_rows_per_page,
    'save-path':           is_api_bool,
    'status':              is_api_bool,
    'tlsa_matching_type':  is_tlsa_matching_type,
    'tlsa_selector':       is_tlsa_selector,
    'tlsa_usage':          is_tlsa_usage,
    'ttl':                 is_ttl,
    'txt':                 is_domain_name,
    'valid':               is_valid,
    'weight':              is_int,
    'zone-type':           is_zone_type,
}
